File: 260-single-number-iii/single-number-iii.cpp
class Solution {
public:
    vector<int> singleNumber(vector<int>& nums) {

        int n = nums.size();

        vector<int> res;

        long long int xr = 0;

        for(int i = 0;i<n;i++){
            xr = xr^nums[i];
        }

        int mask = xr&(-xr);

        int bucket1 = 0, bucket2 = 0;

        for(int i = 0;i<n;i++){
            if(mask & nums[i]){
                bucket1 = bucket1^nums[i];
            }else{
                bucket2 = bucket2^nums[i];
            }
        }

        res.push_back(bucket1);
        res.push_back(bucket2);

        return res;
        
    }
};

File: 260-single-number-iii/single-number-iii.java
class Solution {
    public int[] singleNumber(int[] nums) {

        int n = nums.length;

        int xr = 0;

        for(int x  :nums){
            xr = xr^x;
        }

        int mask = xr & (-xr);

        int[] res = new int[2];

        for(int x : nums){
            if((mask & x)==0){
                res[0] ^= x;
            }
            else{
                res[1] ^=x;
            }
        }

        return res;
        
    }
}